src/chainstate/stacks/db/blocks.rs

```diff
@@ -5272,28 +5273,45 @@ impl StacksChainState {
             _ => false, // unused
         };
 
-        let current_tip =
-            StacksChainState::get_parent_index_block(current_consensus_hash, current_block);
-        let res = match self.with_read_only_clarity_tx(&NULL_BURN_STATE_DB, &current_tip, |conn| {
+        let current_tip = match mempool_admission_check {
+            TransactionAnchorMode::OnChainOnly | TransactionAnchorMode::Any => {
+                StacksChainState::get_parent_index_block(current_consensus_hash, current_block)
+            }
+            TransactionAnchorMode::OffChainOnly => {
+                if let Some(unconfirmed_state) = &self.unconfirmed_state {
+                    if unconfirmed_state.is_readable() {
+                        unconfirmed_state.unconfirmed_chain_tip
+                    } else {
+                        StacksChainState::get_parent_index_block(
+                            current_consensus_hash,
+                            current_block,
+                        )
+                    }
+                } else {
+                    StacksChainState::get_parent_index_block(current_consensus_hash, current_block)
+                }
+            }
+        };
+
+        let res = match self.maybe_read_only_clarity_tx(&NULL_BURN_STATE_DB, &current_tip, |conn| {
             StacksChainState::can_include_tx(conn, &conf, has_microblock_pubk, tx, tx_size)
         }) {
-            Some(r) => r,
-            None => Err(MemPoolRejection::NoSuchChainTip(
+            Ok(Some(r)) => r,
+            Ok(None) | Err(_) => Err(MemPoolRejection::NoSuchChainTip(
```

Comment:
You're right, I should propagate the underlying error here. 

@lgalabru do you think it would be appropriate to
- add an error type to `MemPoolRejection` that takes in type Error. Right now, the only enum that does that in `MemPoolRejection` is `FailedToValidate(Error)`, but that error doesn't make sense here.
- convert the error here to a string and have the error be `MemPoolRejection::Other(stringified_err)`

Reply:
it might be easier to just deal with the error in this block. how does it add value to return an error and then re-match? seems like it takes more mental RAM to keep track of it here, because you have to understand the correspondence between two match blocks instead of just looking at one.

Reply:
this error branch isn't covered by unit tests

Reply:
The rematch also deals with validating a tx against a different type of state, if appropriate. For example, if a tx is unable to be validated against the confirmed chain tip, in some cases we would attempt to validate it against the unconfirmed chain tip. We would want this second validation to happen in the case of an error, hence the reason for the match below.